### PR TITLE
BREAKING: introduce namespaces

### PR DESCRIPTION
**Pull request by dim**

### CODE REVIEW

File: claim.go [hunk 1]
```diff
@@ -19,11 +19,16 @@ func (tc *Claim) Release(_ context.Context) error {
...
+// back to the queue.
 func (tc *Claim) Update(ctx context.Context) error {
+	if err := tc.validate(); err != nil {
+		return err
+	}
+
 	_, err := tc.tx.
 		StmtContext(ctx, tc.update).
-		ExecContext(ctx, tc.Priority, tc.Payload, tc.ID)
+		ExecContext(ctx, tc.Namespace, tc.Priority, tc.Payload, tc.ID)
```
Comment (mxmCherry) [inline claim.go:31]: Hmmm, are you sure you want to have it mutable? Like, flexibility is fine, but does it make much sense?

**Review by mxmCherry — APPROVED** [review]

File: client_test.go [hunk 2]
```diff
@@ -78,11 +78,32 @@ func TestClient_List(t *testing.T) {
...
+	assertEqual(t, tasks, []*TaskDetails{
+		{
+			Task:      Task{ID: task3.ID, Namespace: "baz", Payload: json.RawMessage(`{}`)},
+			CreatedAt: mockNow,
+			UpdatedAt: mockNow,
+		},
+	})
+
+	// bad namespace
+	if _, err := client.List(ctx, WithNamespace("日本国")); err == nil || err.Error() != `namespace "日本国" contains non-ASCII characters` {
```
[the remark this answers is not in the document]
Comment (dim) [reply client_test.go:99]: Die performance reasons I made the column callation C. I was toying with the idea of using ltree but that would have been even more restrictive. The benefit of ^^ is that you can eventually use LIKE queries with wildcard suffixes to work across nested namespaces. That would be much, much slower with utf8 collation.

File: schema.sql [hunk 3]
```diff
@@ -12,31 +12,34 @@ CREATE EXTENSION IF NOT EXISTS pgcrypto;
...
   payload JSONB NOT NULL DEFAULT '{}'
 );
 
-CREATE INDEX IF NOT EXISTS idx_tasks_priority_order ON tasks (priority DESC, updated_at ASC);
+CREATE INDEX IF NOT EXISTS idx_pgpq_tasks_namespace ON pgpq_tasks (namespace ASC);
 
-CREATE INDEX IF NOT EXISTS idx_tasks_created_at ON tasks (created_at ASC);
+CREATE INDEX IF NOT EXISTS idx_pgpq_tasks_created_at ON pgpq_tasks (created_at ASC);
+
+CREATE INDEX IF NOT EXISTS idx_pgpq_tasks_priority_order ON pgpq_tasks (priority DESC, updated_at ASC);
```
[the remark this answers is not in the document]
Comment (dim) [reply schema.sql:28]: Need to think about that...

[the remark this answers is not in the document]
Comment (dim) [reply schema.sql:28]: Probably

File: client.go [hunk 4]
```diff
@@ -82,16 +101,24 @@ func (c *Client) MinCreatedAt(ctx context.Context) (time.Time, error) {
 
 // Push pushes a task into the queue. It may return ErrDuplicateID.
 func (c *Client) Push(ctx context.Context, task *Task) error {
+	if err := task.validate(); err != nil {
+		return err
+	}
+
+	if len(task.Payload) == 0 {
+		task.Payload = json.RawMessage{'{', '}'}
```
[the remark this answers is not in the document]
Comment (mxmCherry) [reply client.go:109]: @dim just realized - maybe `json.RawMessage("null")`, but not too important anyway.

**Pull request merged**